support/refc/runtime.c

```diff
@@ -7,64 +7,83 @@ void missing_ffi() {
   exit(1);
 }
 
-void push_Arglist(Value_Arglist *arglist, Value *arg) {
-  IDRIS2_REFC_VERIFY(arglist->filled < arglist->total,
-                     "unable to add more arguments to arglist");
+static inline Value *idris2_dispatch_closure(Value_Closure *clo) {
+  Value **const xs = clo->args;
+  Value *(*const f)() = clo->f;
 
-  arglist->args[arglist->filled] = newReference(arg);
-  arglist->filled++;
-}
-
-Value *apply_closure(Value *_clos, Value *arg) {
-  // create a new arg list
-  Value_Arglist *oldArgs = ((Value_Closure *)_clos)->arglist;
-  Value_Arglist *newArgs = newArglist(0, oldArgs->total);
-  newArgs->filled = oldArgs->filled + 1;
-  // add argument to new arglist
-  for (int i = 0; i < oldArgs->filled; i++) {
-    newArgs->args[i] = newReference(oldArgs->args[i]);
-  }
-  newArgs->args[oldArgs->filled] = newReference(arg);
-
-  Value_Closure *clos = (Value_Closure *)_clos;
+  switch (clo->arity) {
+  default:
+    fprintf(stderr, "%s: %d params not is suppoted.", __FUNCTION__, clo->arity);
+    exit(1);
 
-  // check if enough arguments exist
-  if (newArgs->filled >= newArgs->total) {
-    fun_ptr_t f = clos->f;
-    while (1) {
-      Value *retVal = f(newArgs);
-      removeReference((Value *)newArgs);
-      if (!retVal || retVal->header.tag != COMPLETE_CLOSURE_TAG) {
-        return retVal;
-      }
-      f = ((Value_Closure *)retVal)->f;
-      newArgs = ((Value_Closure *)retVal)->arglist;
-      newArgs = (Value_Arglist *)newReference((Value *)newArgs);
-      removeReference(retVal);
-    }
+  case 0:
+    return (*f)();
+  case 1:
+    return (*f)(xs[0]);
+  case 2:
+    return (*f)(xs[0], xs[1]);
+  case 3:
+    return (*f)(xs[0], xs[1], xs[2]);
+  case 4:
+    return (*f)(xs[0], xs[1], xs[2], xs[3]);
+  case 5:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4]);
+  case 6:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5]);
+  case 7:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6]);
+  case 8:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6], xs[7]);
+  case 9:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6], xs[7], xs[8]);
+  case 10:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6], xs[7], xs[8],
+                xs[9]);
+  case 11:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6], xs[7], xs[8],
+                xs[9], xs[10]);
+  case 12:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6], xs[7], xs[8],
+                xs[9], xs[10], xs[11]);
+  case 13:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6], xs[7], xs[8],
+                xs[9], xs[10], xs[11], xs[12]);
+  case 14:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6], xs[7], xs[8],
+                xs[9], xs[10], xs[11], xs[12], xs[13]);
+  case 15:
+    return (*f)(xs[0], xs[1], xs[2], xs[3], xs[4], xs[5], xs[6], xs[7], xs[8],
+                xs[9], xs[10], xs[11], xs[12], xs[13], xs[14]);
+  case 16:
```

Comment:
Does this limit us to functions with at most 16 arguments?

Reply:
I may be misremembering, but I believe the old C compilers would only accept about 31 arguments, 31 for C98, maybe even 100 for C99.
Also, considering the stack consumption, I don't think it's a good idea to put everything on the stack. Perhaps some larger functions should be improved to accept arguments as Value* arrays.